### ISSUE TITLE
Equal method returns true but Hashcodes are different 

### ISSUE THREAD

**Issue by SheepRock** [dotnet/runtime#82825]
### Description

From [this StackOverflow question](https://stackoverflow.com/questions/75601660/why-this-simple-struct-returns-different-hash-codes-for-objects-that-are-equal), two instances of a structure with a Datetime property, with the only difference being the Datetime.Kind returns `true` for the `Equal` but have different HashCodes.

### Reproduction Steps

```
var d1 = new DateTime(2023, 2, 28, 0, 0, 0, DateTimeKind.Unspecified);
var d2 = new DateTime(2023, 2, 28, 0, 0, 0, DateTimeKind.Local);

var foo1 = new Foo { d = d1 };
var foo2 = new Foo { d = d2 };

Console.WriteLine($"d1.Equals(d2): {d1.Equals(d2)}");
Console.WriteLine($"d1.GetHashCode() == d2.GetHashCode(): {d1.GetHashCode() == d2.GetHashCode()}");
Console.WriteLine($"foo1.Equals(foo2): {foo1.Equals(foo2)}");
Console.WriteLine($"foo1.GetHashCode() == foo2.GetHashCode(): {foo1.GetHashCode() == foo2.GetHashCode()}");

struct Foo {
    public DateTime d;
}
```

### Expected behavior

```
d1.Equals(d2): True
d1.GetHashCode() == d2.GetHashCode(): True
foo1.Equals(foo2): True
foo1.GetHashCode() == foo2.GetHashCode(): True
```

### Actual behavior

```
d1.Equals(d2): True
d1.GetHashCode() == d2.GetHashCode(): True
foo1.Equals(foo2): True
foo1.GetHashCode() == foo2.GetHashCode(): False
```

### Regression?

No

### Configuration

- .NET 7 (but also previous versions)
- Windows 11
- x64
- Appears to be global

**Comment by tarekgh** [comment]
This looks like a regression from the .NET Framework it worth fixing it.

**Comment by stephentoub** [comment]
> This looks like a regression from the .NET Framework

The repro prints True, True, True, False on both .NET Core and .NET Framework.

**Comment by tarekgh** [comment]
You are right @stephentoub, I misread the code. A quick look at the issue is inside how we calculate the hash code for the value types.  https://github.com/dotnet/runtime/blob/ef0241ed9316bc4343e715daca4d88c1ff90231b/src/coreclr/vm/comutilnative.cpp#L1877. This is not specific to `DateTime`.

**Comment by synek317** [comment]
> The repro prints True, True, True, False on both .NET Core and .NET Framework.

Indeed ([dotnet fiddle](https://dotnetfiddle.net/Em8XA5)).

On the other hand, the question I posted on SO was a simplified case. I encountered this issue when I was using a struct with `DateTime`as a key in a Dictionary. And there is something that I don't understand, because this code:

		var d1 = DateTime.Today;
		var d2 = new DateTime(2023, 2, 28);

		var foo1 = new Foo { d = d1 };
		var foo2 = new Foo { d = d2 }; 

	 	var dict = new Dictionary<Foo, int> { { foo1, 42 } };
		
		Console.WriteLine("Contains foo1: {0}", dict.ContainsKey(foo1));
		Console.WriteLine("Contains foo2: {0}", dict.ContainsKey(foo2));

Prints "True, True" on .NET 4.7.2 ([fiddle](https://dotnetfiddle.net/wrd5QU)) but "True, False" on .NET 7 and earlier .NET core versions ([fiddle](https://dotnetfiddle.net/IHhfsK)).

**Comment by synek317** [comment]
One more update / test case:

I replaced `DateTime` with another struct with overridden `Equals` and `GetHashCode`. As  @tarekgh already stated, it turned out to be not `DateTime`-specific. BUT the `.Contains` test I provided above gives different results on .NET 4.7.2 (It prints "True, False", just like on .NET 7).

[.NET 4.7.2 fiddle](https://dotnetfiddle.net/iplmd4), [.NET 7 fiddle](https://dotnetfiddle.net/1SPjHk)
